src/hmm/topology.cc

```diff
@@ -239,148 +118,178 @@ void Topology::Check() {
-          KALDI_ERR << "Topology::Check(), duplicate transition found.";
-        if (dst_state == k) {  // self_loop...
-          KALDI_ASSERT(entries_[i][j].self_loop_pdf_class != kNoPdf &&
-                       "Nonemitting states cannot have self-loops.");
-        }
-        seen_transition.insert(dst_state);
-        has_trans_in[dst_state] = true;
+  bool has_final_state = false;
+  for (auto const& entry: entries_) {
+    for (fst::StateIterator<fst::StdVectorFst> state_iter(entry);
+         !state_iter.Done(); state_iter.Next()) {
+      StateId state_id = state_iter.Value();
+      if (entry.Final(state_id) != Weight::Zero()) {
+        has_final_state = true;
+
+        // TODO(galv): Figure out whether this is really
+        // necessary. Dan seems to think it is unncessary.
+        if (entry.NumArcs(state_id) > 0)
+          KALDI_ERR << "Topology::Check(), final states must have no outward arcs.";
       }
-      if (j+1 < num_states) {
-        KALDI_ASSERT(tot_prob > 0.0 && "Non-final state must have transitions out."
-                     "(with nonzero probability)");
-        if (fabs(tot_prob - 1.0) > 0.01)
-          KALDI_WARN << "Total probability for state " << j <<
-              " in topology entry is " << tot_prob;
-      } else
-        KALDI_ASSERT(tot_prob == 0.0);
     }
-    // make sure all but start state have input transitions.
-    for (int32 j = 1; j < num_states; j++)
-      if (!has_trans_in[j])
-        KALDI_ERR << "Topology::Check, state "<<(j)<<" has no input transitions.";
-    SortAndUniq(&seen_pdf_classes);
-    if (seen_pdf_classes.front() != 0 ||
-        seen_pdf_classes.back() != static_cast<int32>(seen_pdf_classes.size()) - 1) {
-      KALDI_ERR << "Topology::Check(), pdf_classes are expected to be "
-          "contiguous and start from zero.";
+
+    if (!has_final_state) {
+      KALDI_ERR << "Topology::Check(), must have a final state.";
     }
-  }
-}
 
-bool Topology::IsHmm() const {
-  const std::vector<int32> &phones = GetPhones();
-  KALDI_ASSERT(!phones.empty());
-  for (size_t i = 0; i < phones.size(); i++) {
-    int32 phone = phones[i];
-    const TopologyEntry &entry = TopologyForPhone(phone);
-    for (int32 j = 0; j < static_cast<int32>(entry.size()); j++) {  // for each state...
-      int32 forward_pdf_class = entry[j].forward_pdf_class,
-            self_loop_pdf_class = entry[j].self_loop_pdf_class;
-      if (forward_pdf_class != self_loop_pdf_class)
-        return false;
+    if (entry.Final(entry.Start()) != Weight::Zero())
+      KALDI_ERR << "Topology::Check(), start state must not be a final state.";
+
+    if (entry.Start() != 0) {
+      KALDI_ERR << "Topology::Check(), start state must be 0.";
+    }
+
+
+    // Still TODO:
+    // 1) Check for negative or zero transition probabilities.
+    // 2) Check that there are no pdf classes equal to 0 (this basically means
+    // that all of our transitions must be emitting. Is that okay?)
```

Comment:
Correct, all transitions must be emitting.  Inward arcs to each state do not have to sum to one.  Outward arcs might be expected to sum to one, but I don't know if this merits a warning, since the consequences of unnormalized states are not severe (possible suboptimality in graph search behavior).